sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -283,3 +590,970 @@ def get_lpmode_support(self):
+        This command transfers control from the currently
+        running firmware to a selected firmware that is started. It
+        can be used to switch between firmware versions, or to
+        perform a restart of the currently running firmware.
+        mode:
+        00h = Traffic affecting Reset to Inactive Image.
+        01h = Attempt Hitless Reset to Inactive Image
+        02h = Traffic affecting Reset to Running Image.
+        03h = Attempt Hitless Reset to Running Image
+        """
+        try:
+            self.cdb
+        except AttributeError:
+            self.get_cdb_api()
+        # run module FW (CMD 0109h)
+        starttime = time.time()
+        fw_run_status = self.cdb.run_fw_image(mode)
+        if fw_run_status == 1:
+            logger.info('Module FW run: Success')
+        else:
+            self.cdb.abort_fw_download()
+            logger.info('Module FW run: Fail')
+        elapsedtime = time.time()-starttime
+        logger.info('Module FW run time: %.2f s\n' %elapsedtime)
+
+    def module_fw_commit(self):
+        """
+        The host uses this command to commit the running image
+        so that the module will boot from it on future boots.
+        """
+        try:
+            self.cdb
+        except AttributeError:
+            self.get_cdb_api()
+        # commit module FW (CMD 010Ah)
+        starttime = time.time()
+        fw_commit_status= self.cdb.commit_fw_image()
+        if fw_commit_status == 1:
+            logger.info('Module FW commit: Success')
+        else:
+            self.cdb.abort_fw_download()
+            logger.info('Module FW commit: Fail')
+        elapsedtime = time.time()-starttime
+        logger.info('Module FW commit time: %.2f s\n' %elapsedtime)
+
+    def module_fw_download(self, startLPLsize, maxblocksize, lplonly_flag, autopaging_flag, writelength, imagepath):
+        """
+        This function performs the download of a firmware image to module eeprom
+        It starts CDB download by writing the header of start header size
+        from the designated firmware file to the local payload page 0x9F, with CDB command 0101h.
+
+        Then it repeatedly reads from the given firmware file and write to the payload
+        space advertised from the first step. We use CDB command 0103h to write to the local payload;
+        we use CDB command 0104h to write to the extended paylaod. This step repeats until it reaches
+        end of the firmware file, or the CDB status failed.
+
+        The last step is to complete the firmware upgrade with CDB command 0107h.
+
+        Note that if the download process fails anywhere in the middle, we need to run CDB command 0102h
+        to abort the upgrade before we restart another upgrade process.
+        """
+        try:
+            self.cdb
+        except AttributeError:
+            self.get_cdb_api()
+        # start fw download (CMD 0101h)
+        starttime = time.time()
+        f = open(imagepath, 'rb')
```

Comment:
lets not crash if imagepath is incorrect

Reply:
done